src/main/java/org/eclipsefoundation/marketplace/service/impl/DefaultPromotionService.java

```diff
@@ -0,0 +1,171 @@
+ * can shift outside of code builds once data is modified.
+ * 
+ * @author Martin Lowe
+ *
+ */
+@ApplicationScoped
+public class DefaultPromotionService implements PromotionService {
+	private static final Logger LOGGER = LoggerFactory.getLogger(DefaultPromotionService.class);
+
+	@ConfigProperty(name = MicroprofilePropertyNames.PROMO_SERVE_COUNT, defaultValue = "2")
+	int promoCount;
+	@ConfigProperty(name = MicroprofilePropertyNames.PROMO_WEIGHT_DEFAULT, defaultValue = "1")
+	int defaultWeight;
+
+	@Inject
+	MongoDao dao;
+
+	@Inject
+	DtoFilter<Listing> listingFilter;
+	@Inject
+	CachingService<List<Listing>> listingCache;
+
+	@Inject
+	DtoFilter<Promotion> promotionFilter;
+	@Inject
+	CachingService<List<Promotion>> promoCache;
+
+	// random used for shuffling collections
+	private Random r = new Random();
+
+	@Override
+	public List<Listing> getListingsForPromotions(RequestWrapper wrapper, List<Promotion> promos) {
+		if (promos == null || promos.isEmpty()) {
+			LOGGER.debug("No promotions were passed, returning empty list");
+			return Collections.emptyList();
+		}
+		// create mapping to get a list of specific IDs, and to add context to the call
+		// for caching
+		Map<String, List<String>> adds = new HashMap<>();
+		adds.put("type", Arrays.asList("Listing"));
+		adds.put(UrlParameterNames.IDS.getParameterName(),
+				promos.stream().map(Promotion::getListingId).collect(Collectors.toList()));
+
+		MongoQuery<Listing> q = new MongoQuery<>(null, adds, listingFilter);
+		// retrieve the possible cached object
+		Optional<List<Listing>> cachedResults = listingCache.get("promo|listings", wrapper, adds,
+				() -> StreamHelper.awaitCompletionStage(dao.get(q)));
+		if (!cachedResults.isPresent()) {
+			LOGGER.error("Error while retrieving cached promotion listings");
+			return Collections.emptyList();
+		}
+
+		// return the results as a response
+		return cachedResults.get();
+	}
+
+	@Override
+	public void injectPromotions(RequestWrapper wrapper, List<Listing> listings) {
+		// create an empty promo query to get all promos
+		MongoQuery<Promotion> q = new MongoQuery<>(null, Collections.emptyMap(), promotionFilter);
+		// retrieve the possible cached object
+		Optional<List<Promotion>> cachedResults = promoCache.get("all|promo", wrapper, Collections.emptyMap(),
+				() -> StreamHelper.awaitCompletionStage(dao.get(q)));
+		if (!cachedResults.isPresent() || cachedResults.get().isEmpty()) {
+			LOGGER.debug("Could not find any promotions to inject, returning");
+			return;
+		}
+		// make a copy of the array to not impact cached values
+		List<Promotion> promos = new ArrayList<>(cachedResults.get());
+		List<Promotion> promoHolding = new ArrayList<>(promoCount);
+		LOGGER.debug("Found {} promotions, maximum number to inject {}", promos.size(), promoCount);
+		// check each promotion to see if it should be injected
+		Promotion curr;
+		while ((curr = getWeightedPromotion(promos)) != null && promoHolding.size() <= promoCount) {
```

Comment:
I would build another method "pickNWeightedPromotion(List<Promotion>, int)" that returns n elements from the list rather than this very complicated for loop here

Reply:
The reason for the complicated loop is that currently not all promotions are suitable to be injected based on already present data. If we change this functionality, I could easily see this being a great change, but as of right now I'd be mixing business logic into places it shouldn't be